Goto-contractor PRDC2022 update

Mohannad-Aldughaim opened the pull request:
Hello everyone,

This is an update for --goto-contractor that includes:
-bugfix where esbmc crash when there is an && or || in the program. now it ignores it.
-bugfix where the chosen loop is not the last loop.
-changed assume value when the contractor returns an empty set to 0. assume(0)
-Added regression for last loop and the empty set

Mohannad-Aldughaim committed:
-fixed a bug where the loop chosen was not the last loop.

Mohannad-Aldughaim committed:
-fixed a bug where the last loop wasnt actually the last loop.
-fixed lower bound problem.
-fixed crashes when && || are used.

Mohannad-Aldughaim committed:
-Added regression for empty set and last loop.

Mohannad-Aldughaim committed:
-Added regression for empty set and last loop.

lucasccordeiro commented on src/goto-programs/goto_contractor.cpp:
```diff
@@ -104,39 +103,58 @@ void goto_contractort::parse_intervals(irep_container<expr2t> expr)
   for(auto &function_loop : function_loops)
-    loop = function_loop;
+
+    if(
```
Can you write some documentation here? This piece of code looks a bit confusing.

Mohannad-Aldughaim replied:
This is to find the last loop in the code based on location. Added to documentation.

lucasccordeiro commented on src/goto-programs/goto_contractor.cpp:
```diff
@@ -104,39 +103,58 @@ void goto_contractort::parse_intervals(irep_container<expr2t> expr)
   auto goto_function = goto_functions.function_map.find("c:@F@main")->second;
 
-  for(auto const &var : map.var_map)
+  if(map.isemptyVector)
```
Can you also write some documentation here?

Mohannad-Aldughaim replied:
documentation updated. Also renamed for clarity.

lucasccordeiro commented on src/goto-programs/goto_contractor.cpp:
```diff
@@ -104,39 +103,58 @@ void goto_contractort::parse_intervals(irep_container<expr2t> expr)
+      symbol2tc X = var.second.getSymbol();
+      if(var.second.isIntervalChanged())
+      {
+        /*auto lb = create_value_expr(var.second.getInterval().lb(), int_type2());
```
Shall this commented code be removed?

Mohannad-Aldughaim replied:
removed.

lucasccordeiro commented on src/goto-programs/goto_contractor.cpp:
```diff
@@ -104,39 +103,58 @@ void goto_contractort::parse_intervals(irep_container<expr2t> expr)
-      goto_function.body.destructive_insert(loop_exit, tmp_e2);
+      goto_function.body.destructive_insert(loop_exit, tmp_e);*/
+
+        auto ub = create_value_expr(var.second.getInterval().ub(), int_type2());
```
Can you please add some documentation here?

Mohannad-Aldughaim replied:
Updated the documentation.

lucasccordeiro commented on src/goto-programs/goto_contractor.cpp:
```diff
@@ -164,6 +182,8 @@ void goto_contractort::contractor()
 
   oss << "\n\t- Domains (after): " << X;
   map.update_intervals(X);
+  if(X.is_empty() || X[0].is_degenerated())
```
Please add a comment.

Mohannad-Aldughaim replied:
done. This is to check if interval box is empty set or if the interval is degenerated in the case of a single interval.

lucasccordeiro commented on src/goto-programs/goto_contractor.cpp:
```diff
@@ -192,8 +212,13 @@ goto_contractort::create_constraint_from_expr2t(irep_container<expr2t> expr)
   if(
-    is_arith_expr(expr) || is_constant_number(expr) || is_symbol2t(expr) ||
-    is_notequal2t(expr) || is_equality2t(expr))
+    is_arith_expr(get_base_object(expr)) ||
```
This code looks very weird. Can you move it to a method and explain it?

mikhailramalho replied:
maybe moving the `get_base_object(expr)` calls to before this branch should make it more readable as well.

Mohannad-Aldughaim replied:
Basically a list of unsupported operators. I moved it to a function and performed get_base_object before the checks.

mikhailramalho commented on src/goto-programs/goto_contractor.cpp:
```diff
@@ -104,39 +103,58 @@ void goto_contractort::parse_intervals(irep_container<expr2t> expr)
   {
-    symbol2tc X = var.second.getSymbol();
-    if(var.second.isIntervalChanged())
+    auto cond = create_value_expr(0, int_type2());
```
you can probably replace this with a call to gen_zero.

Mohannad-Aldughaim replied:
done.

mikhailramalho commented on src/goto-programs/goto_contractor.cpp:
```diff
@@ -104,39 +103,58 @@ void goto_contractort::parse_intervals(irep_container<expr2t> expr)
   auto goto_function = goto_functions.function_map.find("c:@F@main")->second;
 
-  for(auto const &var : map.var_map)
+  if(map.isemptyVector)
```
Can we replace `map.isemptyVector` with a `map.var_map.empty()` and delete `map.isemptyVector`?

Mohannad-Aldughaim replied:
done

mikhailramalho commented on src/goto-programs/goto_contractor.cpp:
```diff
@@ -186,8 +205,13 @@ goto_contractort::create_constraint_from_expr2t(irep_container<expr2t> expr)
+    is_modulus2t(get_base_object(expr)) || is_or2t(get_base_object(expr)) ||
+    is_and2t(get_base_object(expr)))
   {
     log_status("Expression is complex, skipping this assert");
```
It should be helpful to print which functions were skipped, you can get the expr name by calling `get_expr_id`.

Also, this looks more like a debug message than a status one, don't you think? I expect to see several of these messages being printed in a row when verifying a program.

Mohannad-Aldughaim replied:
I agree, changed it to include which expression id and changed the type to debug message.

mikhailramalho reviewed (approved):
LGTM, except for a few nitpicks and the documentation that @lucasccordeiro pointed out.

fbrausse commented on src/goto-programs/goto_contractor.h:
```diff
@@ -54,6 +54,7 @@ class CspMap
   static constexpr int NOT_FOUND = -1;
 
   std::map<std::string, vart> var_map;
+  bool isemptyVector = false;
```
In ESBMC there seems to be the convention to not use camelCase for variable or function identifiers. In case you want to keep this member:
1. Could you rename this feld?
2. It is not clear to me which vector it is referring to, could you please add a comment to clarify?

Mohannad-Aldughaim replied:
Renamed to is_empty_set(). Sorry for the confusion; it refers to the result of the contractor (which is an interval vector) whether it is an empty set or not.

Mohannad-Aldughaim committed:
-Renamed isemptyVector to is_empty_vector to follow convention
-Chenged is_empty_vector flag to private and added getter.
-Updated documentation to reflect new changes.
-Changed log message of unsupported operators to debug.
-Updated regression test to finish quickly.

Mohannad-Aldughaim committed:
-Commented ESBMC_GOTO_CONTRACTOR_H in #endif
-updated regression test

mikhailramalho commented on src/goto-programs/goto_contractor.cpp:
```diff
@@ -221,6 +231,25 @@ goto_contractort::create_constraint_from_expr2t(irep_container<expr2t> expr)
   return c;
 }
 
+bool goto_contractort::is_unsupported_operator(irep_container<expr2t> expr)
```
you can rewrite this as `return     is_arith_expr(e) ||
    is_constant_number(e) ||
    is_symbol2t(e) ||
    is_notequal2t(e) ||
    is_equality2t(e) ||
    is_not2t(e) ||
    is_modulus2t(e) ||
    is_or2t(e) ||
    is_and2t(e));`

Mohannad-Aldughaim replied:
Done

mikhailramalho commented on src/goto-programs/goto_contractor.cpp:
```diff
@@ -221,6 +231,25 @@ goto_contractort::create_constraint_from_expr2t(irep_container<expr2t> expr)
 
+bool goto_contractort::is_unsupported_operator(irep_container<expr2t> expr)
+{
+  irep_container<expr2t> e = get_base_object(expr);
```
This is an `expr2tc`.

Mohannad-Aldughaim replied:
Changed

mikhailramalho commented on src/goto-programs/goto_contractor.cpp:
```diff
@@ -98,39 +97,49 @@ void goto_contractort::parse_intervals(irep_container<expr2t> expr)
   auto goto_function = goto_functions.function_map.find("c:@F@main")->second;
 
-  for(auto const &var : map.var_map)
+  if(map.is_empty_set())
```
this is not what I meant, do you only need to check if the var_map is empty here? What do you call `map.var_map.empty()` instead?

Mohannad-Aldughaim replied:
That would be incorrect. Because the check for an empty set is done on the temporarily created IntervalVector object by contractor(). The var_map can have members and the is_empty_set() is true.

mikhailramalho replied:
So, if I got it right: 
1. map holds all constraints.
2. If we didn't add any new constraints during a given pass, we don't add them to the program.
3. If there were constraints added to the map, we add them all to the program.

Is that correct? If so, it might be unsound.

Mohannad-Aldughaim replied:
I don't understand what you mean here 100% though I'll try to answer.

> 1. map holds all constraints.

Yes, it holds each variable that is in the property and its interval. Also holds the property which is our constraint in the Constraint Satisfaction Problem to be used for the contractor. 
> 2.  If we didn't add any new constraints during a given pass, we don't add them to the program.
> 3. If there were constraints added to the map, we add them all to the program.

If the map is empty, we can't apply the contractor; thus, we do not add any `assume()` into the program. If the map is not empty, we apply the contractor and check the result: 

1. If the returned IntervalVector is an empty set, we insert `assume(0)` into the loop. Because it means that the property will never be violated.
2. Otherwise, for each variable, If the intervals have changed, insert assume reflecting the new interval.  

I hope this clarifies the idea. We could also have a small chat online because I would like to hear your thoughts about it.

mikhailramalho commented on src/goto-programs/goto_contractor.h:
```diff
@@ -108,6 +108,11 @@ class CspMap
+    //check if interval box is empty set or if the interval is degenerated
+    // in the case of a single interval
+    if(vector.is_empty()|| vector[0].is_degenerated())
+      is_empty_vector = true;
```
do you really need this boolean variable?

Mohannad-Aldughaim replied:
Yes. it is the only information I need from the contactor() result because the IntervalVector object will no longer exist.

Mohannad-Aldughaim committed:
-updated regression test

Mohannad-Aldughaim committed:
-Changed irep_container<expr2t> to expr2tc
-Also added check if we only have one interval in a vector when updating is_empty_vector.

Mohannad-Aldughaim committed:
-code-style

Mohannad-Aldughaim committed:
-code-style

fbrausse commented on src/goto-programs/goto_contractor.h:
```diff
@@ -120,6 +125,13 @@ class CspMap
+public:
+  bool is_empty_set() const
+  {
+    return is_empty_vector;
```
Thanks for the renaming! However, a vector and a set are quite different types. Would it be possible to agree on one of them?

lucasccordeiro reviewed (approved):
@Mohannad-Aldughaim: could you please answer @mikhailramalho's question? Apart from that, I'm okay with this PR.

lucasccordeiro commented:
@Mohannad-Aldughaim: could you please rebase your branch with the master?

mikhailramalho commented:
Hi Mohannad,

Thank you for your answers, I'll try to run some examples locally to see if
I understood it correctly.

BTW, I already approved the PR, so these questions are only for my own
understanding of the algorithm you wrote, so it shouldn't block the PR.

Em qui., 11 de ago. de 2022 às 12:44, Mohannad ***@***.***>
escreveu:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In src/goto-programs/goto_contractor.cpp
> <https://github.com/esbmc/esbmc/pull/836#discussion_r943640679>:
>
> >
>    auto loop_exit = loop.get_original_loop_exit();
>
>    goto_programt dest;
>
>    auto goto_function = ***@***.***@main")->second;
>
> -  for(auto const &var : map.var_map)
> +  if(map.is_empty_set())
>
> I don't understand what you mean here 100% though I'll try to answer.
>
>
>    1. map holds all constraints.
>
> Yes, it holds each variable that is in the property and its interval. Also
> holds the property which is our constraint in the Constraint Satisfaction
> Problem to be used for the contractor.
>
>
>    1. If we didn't add any new constraints during a given pass, we don't
>    add them to the program.
>    2. If there were constraints added to the map, we add them all to the
>    program.
>
> If the map is empty, we can't apply the contractor; thus, we do not add
> any assume() into the program. If the map is not empty, we apply the
> contractor and check the result:
>
>    1. If the returned IntervalVector is an empty set, we insert assume(0)
>    into the loop. Because it means that the property will never be violated.
>    2. Otherwise, for each variable, If the intervals have changed, insert
>    assume reflecting the new interval.
>
> I hope this clarifies the idea. We could also have a small chat online
> because I would like to hear your thoughts about it.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/esbmc/esbmc/pull/836#discussion_r943640679>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAKEJH5JBJ3IF2EKS7NENJTVYUNWPANCNFSM555N55TA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 

Mikhail Ramalho.
